### PR TITLE
Adapt controller Reference/StateInterfaces to New Way of Exporting (variant support)

### PR DESCRIPTION
**Pull request by mamueluth**
This PR is the fith and last part of multiple breaking down #1240 in smaller changes. For an overview explanation and a lot of comments/discussion, please refer to #1240.
This PR adapts the controller interface to the new way of exporting Reference and StateInterfaces. ReferenceInterfaces now are created as shared_ptr and shared to controller manager as shared_ptr. The memory is still allocated in the controller (old way) but a unordered_map for the interfaces is added.

* Needs https://github.com/ros-controls/ros2_controllers/pull/1103 to be merged as well

**NOTE: Everything is fully backward compatible at this point. At this point variant supports only double at this point.**

### CODE REVIEW

File: controller_interface/include/controller_interface/chainable_controller_interface.hpp
Comment (saikishor) [inline controller_interface/include/controller_interface/chainable_controller_interface.hpp:62]:
```suggestion
  std::vector<std::shared_ptr<const hardware_interface::StateInterface>> export_state_interfaces() final;
```
I think it would make sense to use aliasing with `using`to have something like `hardware_interface::StateInterface::ConstSharedPtr` to `std::shared_ptr<const hardware_interface::StateInterface>`

File: controller_interface/include/controller_interface/chainable_controller_interface.hpp
Comment (saikishor) [inline controller_interface/include/controller_interface/chainable_controller_interface.hpp:65]: Same here as well about the aliasing with `hardware_interface::CommandInterface::SharedPtr`

File: hardware_interface/src/resource_manager.cpp
Comment (saikishor) [inline hardware_interface/src/resource_manager.cpp:1456]: Is it better to modify the `LoanedCommandInterface` and `LoanedStateInterface` to accept the shared_ptr's in the constructor, instead of handling references?

File: controller_interface/include/controller_interface/chainable_controller_interface.hpp
Comment (destogl) [reply controller_interface/include/controller_interface/chainable_controller_interface.hpp:62]: @mamueluth would this work with this `const`?

File: controller_interface/src/chainable_controller_interface.cpp
Comment (destogl) [inline controller_interface/src/chainable_controller_interface.cpp:67]: Do we catch this somewhere? If not, then it is dangerous to throw exceptions as we might break the crash.

File: controller_interface/src/chainable_controller_interface.cpp
Comment (destogl) [inline controller_interface/src/chainable_controller_interface.cpp:111]: If we decided to throw then this comment might go away.

File: hardware_interface/src/resource_manager.cpp
Comment (destogl) [inline hardware_interface/src/resource_manager.cpp:609]: I guess this changes will disappear after #1688 is merged.

File: hardware_interface/src/resource_manager.cpp
Comment (destogl) [reply hardware_interface/src/resource_manager.cpp:1456]: I would say so :)

**Comment by bmagyar** [comment]
I've rebased this PR, you can find a backup of it on https://github.com/StoglRobotics-forks/ros2_control/pull/new/controller_interface_variant_support_base_BAK

File: controller_interface/src/chainable_controller_interface.cpp
Comment (bmagyar) [reply controller_interface/src/chainable_controller_interface.cpp:111]:
```suggestion
``` 
correct! removing

File: hardware_interface/include/hardware_interface/resource_manager.hpp
Comment (bmagyar) [inline hardware_interface/include/hardware_interface/resource_manager.hpp:144]:
```suggestion
    const std::string & controller_name, std::vector<StateInterface::SharedPtr> & interfaces);
```

File: hardware_interface/src/resource_manager.cpp
Comment (bmagyar) [inline hardware_interface/src/resource_manager.cpp:1244]:
```suggestion
  const std::string & controller_name, std::vector<StateInterface::SharedPtr> & interfaces)
```

File: controller_interface/src/chainable_controller_interface.cpp
Comment (bmagyar) [reply controller_interface/src/chainable_controller_interface.cpp:67]: We are catching them, yes.

File: controller_interface/include/controller_interface/chainable_controller_interface.hpp
Comment (bmagyar) [reply controller_interface/include/controller_interface/chainable_controller_interface.hpp]: That's a different PR now

File: controller_interface/include/controller_interface/chainable_controller_interface.hpp
Comment (bmagyar) [reply controller_interface/include/controller_interface/chainable_controller_interface.hpp]: done

File: hardware_interface/src/resource_manager.cpp
Comment (bmagyar) [reply hardware_interface/src/resource_manager.cpp]: Done but only as a syntactical sugar thing. I've made the constructor do the dereferencing... The whole dereferencing to take a reference of a shared ptr is a bad idea and I think we should refactor that asap but would not hold this PR up. I've added deprecation notices to the old constructors to remind us of the sloppy stuff.

I took a stab at removing the reference entirely and using the shared_ptr but it would require rewriting all the controller tests which I'm not available to do at the moment. It may also need a little more thinking on the design of this API...

**Review by saikishor — APPROVED** [review]
Overall the PR looks good to me. Thanks for the great work @mamueluth @bmagyar 👏🏾 

I've found minor things to be addressed apart from the suggestion from this review, but I can open up a following PR to address them.

**Comment by bmagyar** [comment]
@saikishor separate PR please

**Comment by saikishor** [comment]
> @saikishor separate PR please

Sure, I will do that :)

File: controller_manager/src/controller_manager.cpp
Comment (destogl) [inline controller_manager/src/controller_manager.cpp:782]: Let's add Test for this!

**Review by destogl — APPROVED** [review]

**Pull request merged**
